From 9f2e286b0c6da163ee6196c34be115260b5008e6 Mon Sep 17 00:00:00 2001
From: Colin Watson <cjwatson@debian.org>
Date: Mon, 11 May 2015 15:18:37 +0100
Subject: [PATCH] rts/Linker.c: add some more PPC relocations (#10402)

This implements R_PPC_PLTREL24, R_PPC_REL16_LO, R_PPC_REL16_HI, and
R_PPC_REL16_HA, emitted by other parts of the current toolchain.
---
 rts/Linker.c | 13 +++++++++++++
 1 file changed, 13 insertions(+)

Index: ghc-7.8.4/rts/Linker.c
===================================================================
--- ghc-7.8.4.orig/rts/Linker.c	2015-05-13 21:04:36.925078989 +0200
+++ ghc-7.8.4/rts/Linker.c	2015-05-13 21:04:36.921078903 +0200
@@ -5622,6 +5622,7 @@
             break;

          case R_PPC_REL24:
+         case R_PPC_PLTREL24:
             delta = value - P;

             if( delta << 6 >> 6 != delta )
@@ -5641,6 +5642,18 @@
             *(Elf_Word *) P = (*(Elf_Word *) P & 0xfc000003)
                                           | (delta & 0x3fffffc);
             break;
+
+         case R_PPC_REL16_LO:
+            *(Elf32_Half*) P = value - P;
+            break;
+
+         case R_PPC_REL16_HI:
+            *(Elf32_Half*) P = (value - P) >> 16;
+            break;
+
+         case R_PPC_REL16_HA:
+            *(Elf32_Half*) P = (value + 0x8000 - P) >> 16;
+            break;
 #        endif

 #if x86_64_HOST_ARCH
